coreapi/retina_net.py

```diff
@@ -0,0 +1,155 @@
+    """
+
+    def __init__(self, down_scale_factor=1.0):
+
+        self.down_scale_factor = down_scale_factor
+
+    @staticmethod
+    def _pad_input_image(img, max_steps):
+        """pad image to suitable shape"""
+        img_h, img_w, _ = img.shape
+
+        img_pad_h = 0
+        if img_h % max_steps > 0:
+            img_pad_h = max_steps - img_h % max_steps
+
+        img_pad_w = 0
+        if img_w % max_steps > 0:
+            img_pad_w = max_steps - img_w % max_steps
+
+        padd_val = np.mean(img, axis=(0, 1)).astype(np.uint8)
+        img = cv2.copyMakeBorder(img, 0, img_pad_h, 0, img_pad_w,
+                                 cv2.BORDER_CONSTANT, value=padd_val.tolist())
+        pad_params = [img_h, img_w, img_pad_h, img_pad_w]
+
+        return img, pad_params
+
+    @staticmethod
+    def _recover_pad_output(outputs, pad_params):
+        """recover the padded output effect"""
+        img_h, img_w, img_pad_h, img_pad_w = pad_params
+        recover_xy = np.reshape(outputs[:, :14], [-1, 7, 2]) * \
+            [(img_pad_w + img_w) / img_w, (img_pad_h + img_h) / img_h]
+        outputs[:, :14] = np.reshape(recover_xy, [-1, 14])
+        return outputs
+
+    @staticmethod
+    def _draw_bbox_landm(img, ann, img_height, img_width):
+        """draw bboxes and landmarks"""
+     # bbox
+        x1, y1, x2, y2 = int(ann[0] * img_width), int(ann[1] * img_height), \
+            int(ann[2] * img_width), int(ann[3] * img_height)
+        cv2.rectangle(img, (x1, y1), (x2, y2), (0, 255, 0), 2)
+
+       # confidence
+        text = "{:.4f}".format(ann[15])
+        cv2.putText(img, text, (int(ann[0] * img_width), int(ann[1] * img_height)),
+                    cv2.FONT_HERSHEY_DUPLEX, 0.5, (255, 255, 255))
+
+        # landmark
+        if ann[14] > 0:
+            cv2.circle(img, (int(ann[4] * img_width),
+                             int(ann[5] * img_height)), 1, (255, 255, 0), 2)
+            cv2.circle(img, (int(ann[6] * img_width),
+                             int(ann[7] * img_height)), 1, (0, 255, 255), 2)
+            cv2.circle(img, (int(ann[8] * img_width),
+                             int(ann[9] * img_height)), 1, (255, 0, 0), 2)
+            cv2.circle(img, (int(ann[10] * img_width),
+                             int(ann[11] * img_height)), 1, (0, 100, 255), 2)
+            cv2.circle(img, (int(ann[12] * img_width),
+                             int(ann[13] * img_height)), 1, (255, 0, 100), 2)
+
+    @staticmethod
+    def _draw_anchor(img, prior, img_height, img_width):
+        """draw anchors"""
+        x1 = int(prior[0] * img_width - prior[2] * img_width / 2)
+        y1 = int(prior[1] * img_height - prior[3] * img_height / 2)
+        x2 = int(prior[0] * img_width + prior[2] * img_width / 2)
+        y2 = int(prior[1] * img_height + prior[3] * img_height / 2)
+        cv2.rectangle(img, (x1, y1), (x2, y2), (0, 0, 0), 1)
+
+    @staticmethod
```

Comment:
can you put all these utility function in a separate module?

Reply:
Sure!